### PR TITLE
Remove old File Upload article + improve the new one

### PR DESCRIPTION
**Pull request by wouterj**
| Q | A |
| --- | --- |
| Doc fix? | yes |
| New docs? | yes |
| Applies to | 2.3+ |
| Fixed tickets | #5375 |

The old file upload article wasn't good, a new one was written by @javiereguiluz but the old one remained online. This PR removes the old one and documents the missing bits in the new article.

### CODE REVIEW

**Comment by WouterJ** [comment]
Btw, why do we use an Entity property to store both the uploaded file and file path? Wouldn't it be better to use an unmapped field?

File: cookbook/controller/upload_file.rst [hunk 1]
```diff
@@ -164,15 +187,203 @@ There are some important things to consider in the code of the above controller:
...
+
+        // ...
+    }
+
+Using a Doctrine Listener
+-------------------------
+
+If you are using Doctrine to store the Product entity, you can create a
+:doc:`Doctrine listener </cookbook/doctrine/event_listeners_subscribers>` to
+automatically upload the file when persisting the entity::
```
Comment (WouterJ) [inline cookbook/controller/upload_file.rst]: I wondering if we should show code + config, or just link to the other article and just mentioning this as a little tip.

File: cookbook/controller/upload_file.rst [hunk 2]
```diff
@@ -164,15 +187,203 @@ There are some important things to consider in the code of the above controller:
...
+    .. code-block:: html+php
+
+        <a href="<?php echo $view['assets']->getUrl('uploads/brochures/'.$product->getBrochure()) ?>">
+            View brochure (PDF)
+        </a>
+
+Creating an Uploader Service
+----------------------------
+
+To avoid logic in controllers, making them big, you can extract the upload
```
Comment (WouterJ) [inline cookbook/controller/upload_file.rst]: Not sure if I really like this service, it's small doesn't do much and doesn't seem to make things simpler in any way. However, it felt like a nice step between a big controller and using a Doctrine listener.

**Comment by Raistlfiren** [comment]
@WouterJ, if I might make an addition as well.... It might be good to include how to handle files when editing them. For example, if I have a simple form like you included in the example, then I would need to either set data_class to null on the brochure field or convert the string to a File object. Otherwise, the form will error out and say - "The form's view data is expected to be an instance of class Symfony\Component\HttpFoundation\File\File, but is a(n) string". Maybe creating the file object is better in a data transformer, not sure. Thoughts?

**Comment by xabbuh** [comment]
> Btw, why do we use an Entity property to store both the uploaded file and file path? Wouldn't it be better to use an unmapped field?

I think this was done to always trigger the lifecycle callback methods. If only the file path were mapped to the database, lifecycle callbacks would never be triggered if only the uploaded file was changed (Doctrine would not treat the entity as changed at all).

However, I tend to say that we should rather move the whole file uploding logic into a DTO and not pollute the entity with logic that doesn't belong there.

File: cookbook/controller/upload_file.rst [hunk 3]
```diff
@@ -146,16 +159,26 @@ Finally, you need to update the code of the controller that handles the form::
...
 
+Now, create the ``brochures_directory`` parameter that was used in the
+controller to specify the directory in which the brochures should be stored:
+
+.. code-block:: yaml
+
+    # app/config/parameters.yml
+    parameters:
+        # ...
+        brochures_directory: '%kernel.root_dir%/../web/uploads/brochures'
```
Comment (weaverryan) [inline cookbook/controller/upload_file.rst]: This makes more sense in `config.yml` - it's not a value that would normally need to be changed on a machine-by-machine basis, and devs wouldn't/shouldn't need to worry about configuring it.

File: cookbook/controller/upload_file.rst (hunk 1 above)
Comment (weaverryan) [inline cookbook/controller/upload_file.rst]: I like how you've handled it here: with code+config (it's really not too long).

File: cookbook/controller/upload_file.rst [hunk 4]
```diff
@@ -164,15 +187,203 @@ There are some important things to consider in the code of the above controller:
...
+            // upload only works for Product entities
+            if (!$entity instanceof Product) {
+                return;
+            }
+
+            $file = $entity->getBrochure();
+            $fileName = $this->uploader->upload($file);
+
+            $entity->setBrochure($fileName);
+        }
```
Comment (weaverryan) [inline cookbook/controller/upload_file.rst]: Can we add the preUpdate too? That would be more useable

File: cookbook/controller/upload_file.rst (hunk 3 above)
Comment (weaverryan) [inline cookbook/controller/upload_file.rst]: And actually, we're dependent on it later when we make the paths to it in the Twig template.

**Comment by weaverryan** [comment]
@WouterJ I added 2 small comments, but :+1: from me after that. Sorry for the long delay - I'm also excited to remove the old, terrible article.

**Comment by Koc** [comment]
@WouterJ please add info about uploading and validating multiple files, see #5603 request

**Comment by Koc** [comment]
@weaverryan can we create redirect from old url to the new article?

**Comment by xabbuh** [comment]
@Koc That's done with the entry in the `redirection_map` file.

**Comment by WouterJ** [comment]
I've updated the article. Should be ready to merge now

File: cookbook/controller/upload_file.rst [hunk 5]
```diff
@@ -164,15 +188,220 @@ There are some important things to consider in the code of the above controller:
...
+        {
+            $fileName = md5(uniqid()).'.'.$file->guessExtension();
+
+            $file->move($this->targetDir, $fileName);
+
+            return $fileName;
+        }
+    }
+
+This class can be registered as a service in the service container:
```
Comment (javiereguiluz) [inline cookbook/controller/upload_file.rst]: `This class can be registered as a service in the service container:`

->

`Then, define a service for this class:` ?

File: cookbook/controller/upload_file.rst [hunk 6]
```diff
@@ -164,15 +188,220 @@ There are some important things to consider in the code of the above controller:
...
+            if (!$file instanceof UploadedFile) {
+                return;
+            }
+
+            $fileName = $this->uploader->upload($file);
+            $entity->setBrochure($fileName);
+        }
+    }
+
+Now, register this class as a Doctrine listener:
```
Comment (javiereguiluz) [inline cookbook/controller/upload_file.rst]: I have a question about this. In the popular VichUploaderBundle, they warn you about uploading files in entities which haven't modified any other property. In that case, it won't work and the file will be lost. That's why you need to fake some changes (e.g. updating the `updatedtAt` property). Would this be a problem here too?

Comment (WouterJ) [inline cookbook/controller/upload_file.rst]: Hmm, will have to look into this. I think it might is (as doctrine doesn't discover the entity changed)

Comment (xabbuh) [inline cookbook/controller/upload_file.rst]: If I understand the code correctly, the file name will always change when uploaded thus leading to changes in the entity which means that Doctrine is able to detect the change.

Comment (WouterJ) [inline cookbook/controller/upload_file.rst]: Created a test project and this is not a problem with this code.

**Comment by WouterJ** [comment]
While testing this code in an application I found a few bugs which I fixed in f6068c9. Also, I strumbled onto issues creating an edit form as it needs a `File` instance. To avoid problems for readers, I've added two little tip boxes in 0c54b8a.

It's now even more ready to merge than it was this afternoon :)

**Comment by javiereguiluz** [comment]
@WouterJ amazing! Thanks for testing everything thoroughly in a real project 🙇

File: cookbook/controller/upload_file.rst [hunk 7]
```diff
@@ -164,15 +188,268 @@ There are some important things to consider in the code of the above controller:
    that information. That's why it's always better to generate a unique name and
    use the :method:`Symfony\\Component\\HttpFoundation\\File\\UploadedFile::guessExtension`
    method to let Symfony guess the right extension according to the file MIME type;
-#. The ``UploadedFile`` class also provides a :method:`Symfony\\Component\\HttpFoundation\\File\\UploadedFile::move`
-   method to store the file in its intended directory. Defining this directory
-   path as an application configuration option is considered a good practice that
-   simplifies the code: ``$this->container->getParameter('brochures_dir')``.
 
-You can now use the following code to link to the PDF brochure of an product:
+You can use the following code to link to the PDF brochure of an product:
```
Comment (xabbuh) [inline cookbook/controller/upload_file.rst]: a product

**Comment by xabbuh** [comment]
👍 I like the changes you made here.

However, for the future I think we can improve things even more by introducing a DTO that would be used to build the form and doing the file upload. That way we would not need to have the hacks being present currently.

**Comment by weaverryan** [comment]
I left a minor comment, but let's merge this :) 👍

**Comment by WouterJ** [comment]
Updated again. @xabbuh using a DTO is exactly what I thought when improving this article.

@weaverryan I can't find your minor comment.

**Comment by javiereguiluz** [comment]
@WouterJ you say "updated again" but I can't see any new commits 😕

**Comment by WouterJ** [comment]
@javiereguiluz as it was a very small typo fix, I merged it into the previous commit: https://github.com/symfony/symfony-docs/pull/6040/commits/94e21ff80bff77063efc0924ff84331c2ad7250c#diff-bcae3aaf74848d919ee531cad0d95547R192

**Comment by xabbuh** [comment]
Thank you Wouter.

**Pull request closed without merging**